ibis/backends/sql/dialects.py

```diff
@@ -135,6 +134,64 @@ class Generator(Hive.Generator):
             sge.Interval: _interval_with_precision,
         }
 
+        def struct_sql(self, expression: sge.Struct) -> str:
+            from sqlglot.optimizer.annotate_types import annotate_types
+
+            expression = annotate_types(expression)
+
+            values = []
+            schema = []
+
+            for e in expression.expressions:
+                if isinstance(e, sge.PropertyEQ):
+                    e = sge.alias_(e.expression, e.this)
+                # named structs
+                if isinstance(e, sge.Alias):
+                    if e.type and e.type.is_type(sge.DataType.Type.UNKNOWN):
+                        self.unsupported(
+                            "Cannot convert untyped key-value definitions (try annotate_types)."
+                        )
+                    else:
+                        schema.append(f"{self.sql(e, 'alias')} {self.sql(e.type)}")
+                    values.append(self.sql(e, "this"))
+                else:
+                    values.append(self.sql(e))
+
+            if not (size := len(expression.expressions)) or len(schema) != size:
+                return self.func("ROW", *values)
+            return f"CAST(ROW({', '.join(values)}) AS ROW({', '.join(schema)}))"
+
+        def array_sql(self, expression: sge.Array) -> str:
+            # workaround for the time being because you cannot construct an array of named
+            # STRUCTs directly from the ARRAY[] constructor
+            # https://issues.apache.org/jira/browse/FLINK-34898
+            from sqlglot.optimizer.annotate_types import annotate_types
+
+            expression = annotate_types(expression)
+            first_arg = seq_get(expression.expressions, 0)
+            if isinstance(first_arg, sge.Struct):
+                # it's an array of structs
+                named_structs = False
+                for arg in expression.expressions:
+                    for e in arg.expressions:
+                        if isinstance(e, sge.Alias):
+                            named_structs = True
+                # get rid of aliasing because we want to compile this as CAST instead
+                args = deepcopy(expression.expressions)
+                if named_structs:
+                    for arg in args:
+                        arg.set("expressions", [e.this for e in arg.expressions])
+
+                format_values = ", ".join([self.sql(arg) for arg in args])
```

Comment:
```suggestion
            first_arg = seq_get(expression.expressions, 0).copy()
            if isinstance(first_arg, sge.Struct):
                for arg in expression.expressions:
                    arg.set("expressions", [e.unalias() for e in arg.expressions])

                format_values = ", ".join(
                    [self.sql(arg) for arg in expression.expressions]
                )
```

there's an `unalias` method on sqlglot objects that I think lets us simplify this a fair bit (but I might not be covering some corner cases you've run across)

Reply:
Thanks for the pointer - I like this solution and it definitely makes the code much cleaner! I'm tempted to modify on a deepcopy of  `expression.expression` to avoid unexpected consequences with passing this around...